### DISCUSSION TITLE
How to force node to be last stop on route

### DISCUSSION THREAD

**Discussion by satyenataagyaa** [google/or-tools#3528] [Routing (and legacy CP) questions] [answered]
I am attempting to model a VRP where I can force some nodes to be the last to be visited on their respective routes.

1. I start with the "Simple Vehicles Routing Problem" (https://developers.google.com/optimization/routing/vrp)
2. For simplicity, I reduce the number of nodes to the first 8 in the sample, and the number of vehicles to 2.
3. The solver gives me the following result:

```
Objective: 158304
Route for vehicle 0:
 0 ->  8 ->  6 ->  2 ->  5 -> 0
Distance of the route: 1552m

Route for vehicle 1:
 0 ->  1 ->  4 ->  3 ->  7 -> 0
Distance of the route: 1552m

Maximum of the route distances: 1552m
```

4. To force nodes 1 & 8 to be the last stops on their respective routes, I add constraints in the following manner:
```
    solver = routing.solver()
    for idx in [1, 8]:
        node_idx = manager.NodeToIndex(idx)
        solver.Add(routing.NextVar(node_idx) == 0)
```
5. Unfortunately, this results in:
```
No solution found !
```

Apparently, I am not setting the constraints correctly. I request your help with how I can go about doing this.

(The entire program is attached.)
[vrp.py.txt](https://github.com/google/or-tools/files/9897130/vrp.py.txt)

Thank you,
Satyen

**Comment by jmarca** [comment, marked as answer]
I think it is incorrect to assume the last index of a route is 0



James

**Reply by satyenataagyaa** [reply]
My understanding is that since I have created the RoutingIndexManager using:
```
manager = pywrapcp.RoutingIndexManager(len(data['distance_matrix']), data['num_vehicles'], data['depot'])
```
all routes will begin and end at the depot (index 0). This is what I see when I run the program without any constraints.

My constraints are meant to ask the solver to ensure that the node following nodes 1 and 8 has the index 0. Except I don't know if the way I'm asking the solver is correct.

If I should not assume the last index will be that of the depot, can you suggest another strategy to achieve my objective -- to tell the solver to place certain nodes only at the end of their respective routes?

**Reply by Regista6** [reply]
Maybe this can be of help to you.

https://developers.google.com/optimization/routing/routing_tasks#setting-start-and-end-locations-for-routes

**Reply by satyenataagyaa** [reply]
Thank you for the suggestion.

The "Setting start and end locations" approach could work for my toy example. In my real problem, however, the nodes I am trying to set as the last nodes are also "optional," and may be left out of the assignment/solution. The actual requirement is to force specific nodes to be the last on their routes iff they are included in the solution. Do you think "Setting end locations" can help with this extended requirement?

**Reply by Regista6** [reply]
Maybe I am misunderstanding but if you can somehow set such nodes to be just after the depot(0) and then iterate over the route back to front... is that feasible?

**Reply by satyenataagyaa** [reply]
This is a great suggestion! I will give it a try.

**Reply by Regista6** [reply]
Maybe something like this can work.

```py
# New Dimension
    def new_callback(from_index, to_index):
        return 1

    new_callback_index = routing.RegisterTransitCallback(new_callback)

    new_dimension_name = 'new_dimension'
    routing.AddDimension(
        new_callback_index,
        0,  # no slack
        3000,  #
        True,  # start cumul to zero
        new_dimension_name)
    new_dimension = routing.GetDimensionOrDie(new_dimension_name)

    for nodes in [1, 8]:
      node_idx = manager.NodeToIndex(nodes)
      new_dimension.CumulVar(node_idx).SetRange(1,1)

```

**Reply by satyenataagyaa** [reply]
This approach worked nicely for me. I really appreciate your time and advice. Thank you!

**Comment by jmarca** [comment]
It is my understanding that the correct usage is

routing.Start(v)

and

routing.End(v)

where v is the number of the vehicle.

See:

https://developers.google.com/optimization/reference/python/constraint_solver/pywrapcp#end

Using raw integers is generally going to cause problems, even if it
does work.

That said, the root cause of your error might be something else.  Is
it always true that the last nodes are those nodes?  Maybe the initial
heuristic cannot get there, and must start somewhere else and then get
there?




On Sun, Oct 30, 2022 at 07:03:52PM -0700, satyenataagyaa wrote:
> My understanding is that since I have created the RoutingIndexManager using:
> ```
> manager = pywrapcp.RoutingIndexManager(len(data['distance_matrix']), data['num_vehicles'], data['depot'])
> ```
> all routes will begin and end at the depot (index 0). This is what I see when I run the program without any constraints.
>
> My constraints are meant to ask the solver to ensure that the node following nodes 1 and 8 has the index 0. Except I don't know if the way I'm asking the solver is correct.
>
> If I should not assume the last index will be that of the depot, can you suggest another strategy to achieve my objective -- to tell the solver to place certain nodes only at the end of their respective routes?
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/google/or-tools/discussions/3528#discussioncomment-4014750
> You are receiving this because you commented.
>
> Message ID: ***@***.***>

--

James E. Marca
Activimetrics LLC

**Reply by satyenataagyaa** [reply]
I tried asking the solver to ensure nodes 5 and 7 are last nodes (one before the end, which is the depot). These are the nodes picked by the solver when I run without any constraints. I still get "No solution found!"

Are there any logs that provide a view of what the solver is doing to solve the problem?

**Comment by paultrow** [comment]
Take a look at
https://developers.google.com/optimization/routing/routing_tasks#setting-start-and-end-locations-for-routes

**Comment by jmarca** [comment]
At least one bug is here:

```py
    for idx in [1, 8]:
        node_idx = manager.NodeToIndex(idx)
        solver.Add(routing.NextVar(node_idx) == 0)
```

As I said, using 0 for the depot node is nonsensical.

You must set it with End(veh), but you have two vehicles.

So to do it this way, you'd have to loop over the vehicles, and then set a conditional
constraint.

```
Loop =>  for veh in vehicles:
Condition => vehicle=veh,
constraint => NextVar(node) == routing.End(veh)
```

That is super convoluted, and conditional constraints are not well
documented, but they do work.

You might be able to hack around it using a counting dimension of
sorts, and saying that your possible last nodes must equal the
CumulVar(routing.End()) values, but again, I can only see how to do it
using conditional constraint conditioned on whether the vehicle serves
the node or not.

James


On Mon, Oct 31, 2022 at 10:46:15AM -0700, satyenataagyaa wrote:
> Thank you for the suggestion.
>
> The "Setting start and end locations" approach could work for my toy example. In my real problem, however, the nodes I am trying to set as the last nodes are also "optional," and may be left out of the assignment/solution. The actual requirement is to force specific nodes to be the last on their routes iff they are included in the solution. Do you think "Setting end locations" can help with this extended requirement?
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/google/or-tools/discussions/3528#discussioncomment-4021109
> You are receiving this because you commented.
>
> Message ID: ***@***.***>

--

James E. Marca
Activimetrics LLC

**Reply by satyenataagyaa** [reply]
Thank you for pointing out the problem in my code.

I am using the advice from @Regista6 -- to constrain the nodes of interest to be first nodes (instead of last), and invert the order of nodes after they are assigned to a route. This is working for me.

I am sure I will need to use the information you have shared about Conditional Constraints in the near future, and I thank you for your time.
